### PR TITLE
106 synthetic datasets

### PR DESCRIPTION
**Pull request by philippkolbe**

### CODE REVIEW

File: src/neural_bandits/benchmark/datasets/synthetic.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import torch
+
+from neural_bandits.benchmark.datasets.abstract_dataset import AbstractDataset
+
+
+class SyntheticDataset(AbstractDataset[torch.Tensor]):
+    """A binary classification synthetic dataset for bandit algorithms.
+
+    Subclasses should implement the `phi` method to define the feature mapping.
+    """
```
Comment (Rob2U) [inline src/neural_bandits/benchmark/datasets/synthetic.py:12]: I think a short comment about how `phi` is used, the weights are sampled etc. wouldn't hurt here.

File: src/neural_bandits/benchmark/datasets/synthetic.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        """Return the contextualized actions and rewards."""
+        contextualized_actions = self.contextualizer(self.X[idx].unsqueeze(0)).squeeze(0)
+        rewards = torch.tensor(
+            [self.reward(idx, action) for action in range(self.num_actions)],
+            dtype=torch.float32,
+        )
+        return contextualized_actions, rewards
+
+    def reward(self, idx: int, action: int) -> float:
+        """Return the reward for a given index and action."""
```
Comment (Rob2U) [inline src/neural_bandits/benchmark/datasets/synthetic.py:49]: You only do classification here. What is your opinion on doing a dataset with multiple arms where the reward of each arm is a function of the input context. I think this is the standard version used, e.g. in [this paper](https://arxiv.org/pdf/1911.04462). 
This would translate to having multiple `self.theta_opt` one for each arm representing the coefficients of the different basis functions.
